### ISSUE TITLE
Custom metamodel for reactions

### ISSUE THREAD

**Issue by HeikoKlare** [vitruv-tools/Vitruv#306]
We currently use the metamodel for Reactions generated by the Xtext compiler from the specified grammar. We perform some postprocessing in the `ReactionsLanguageXtext2EcorePostProcessor` (adding some opposite references).
Unfortunately, the post processor interface `IXtext2EcorePostProcessor` is deprected, which is probably to avoid problems due to the post processing making the metamodel incompatible to the generated parser if not performed properly.

Before the interface gets removed and postprocessing becomes unsupported, we should switch to a manually developed metamodel that is referred to in the Xtext grammar. This may even make some workarounds in the grammar obsolete (using fragments), which were only added to enforce the generation of a metamodel with a specific structure.
We may simply use the currently generated metamodel and reference it in the language instead of generating it. Afterwards, we can also perform some refactoring of the metamodel to improve its quality.

**Comment by jGleitz** [comment]
Note that you don’t have to create the whole metamodel manually. You can model the parts you want to exert control over manually, and have XText derive a model for the rest.

**Comment by HeikoKlare** [comment]
That's a valuable hint! Since our adaptations only affect the top-level structure (especially reactions segments), we should probably only define that overall structure in an explicit metamodel and let Xtext still derive the metamodel for the reactions and routines implementations.

**Comment by HeikoKlare** [comment]
This is more complicated then I expected, because the Reactions language metamodel has to reference metaclasses from the MirBase language metamodel (e.g. `DomainReference`). In addition, the Reactions language is derived from the MirBase language to reuse its rules, scoping and validation. The Xtext generator, however, is not able to match the metaclass for the inherited rule `DomainReference` with the metaclass referenced by the metaclass predefined in the metamodel (in sich case, the `ReactionsSegment`). This is simply because the two `DomainReference`s for the rule and for the predefined metamodel are loaded into different metamodels and matching whether an element defined in a rule already exists in the predefined metamodel is done by a reference equality check. Thus, the language generator errors because it things there is "an element with the same name but semantically different".

Long story short: Inheriting from a language and using a predefined metamodel that references elements from the inherited language is currently not supported (or implemented correctly) by Xtext. This makes it more difficult for us to replace the postprocessor.

**Comment by jGleitz** [comment]
MirBase introduces a lot of complexity, from my point of view. So much so that I think we should consider whether the code-deduplication that MirBase brings is worth its cost.

**Comment by HeikoKlare** [comment]
I agree. We could think about MirBase only providing a metamodel with common elements and, for example, some reuable scoping or validation logic for them, but without placing them into an Xtext-based language. Then, the concrete languages can use that common metamodel and the logic for them without the necessity to be derived from a common language.

**Comment by jGleitz** [comment]
Sounds good. That’s roughly the pattern I followed when I implemented for the shared logic for adding domain dependencies. So if we went down this route, we could maybe even reuse some bits for the Commonalities Language.

**Comment by HeikoKlare** [comment]
What makes this unnecessarily complicated is the problem that within Eclipse (for validation in the Xtext editor) and within standalone mode (during the MWE2 execution) the Ecore metamodel is resolved with different kinds of URIs (once with its namespace URI, once with a platform URI), such that with either kind of URI used in the language elements metamodel, the references to elements from the Ecore metamodel are not properly matched in either the editor or the MWE2 execution, such that either of them throws errors.
The Commonalities language does not suffer from that because it has an own representation of Ecore metamodel elements. Thus, we can do the following:
1. We could adopt that representation for the other languages, but this would require quite some adaptation of the languages code.
2. We could adapt the Xtext implementation and hope for a PR being accepted.
3. We could find out that the problem is at my side and someone else finds a solution.

The problem can be reproduced easily: Create a metamodel with a single element referencing some element from the Ecore metamodel (such as an `EReference`). Then create a new Xtext project for that Ecore metamodel and add an according parser rule for that single element. The MWE2 workflow will fail with an error, because it cannot match the `EReference` of the metamodel with the `EReference` referenced in the language due to identity comparison of `EReference` loaded in different resource sets.

**Comment by jGleitz** [comment]
Which kinds of URIs are used by the editor / mwe2, respectively? If the editor uses `platform:/resource` URIs but those cannot be resolved in the MWE2 workflow, we can probably just register the Ecore metamodel at the start of the MWE2 workflow.

**Comment by HeikoKlare** [comment]
I think the editor uses platform URIs while MWE2 uses namespace URIs. More precisely, when the metamodel references the Ecore metamodel by its namespace URI, the editor contains errors but the MWE2 workflow succeeds, and when the metamodel references the Ecore metamodel by its platform URI, the editor is error-free but the MWE2 workflow fails.
I tried registering either kind of URI in the MWE2 workflow, but it did not make a difference.

**Comment by jGleitz** [comment]
I was able to reproduce the problem, but not to find a solution. If I create an Ecore model referencing the Ecore meta-metamodel, I get the message

> Cannot find compatible feature referenceReference in sealed EClass Example from imported package https://example.com/mymodel: The type 'EReference' used in the reference 'referenceReference' is inconsistent. Probably this is due to an unsupported kind of metamodel hierarchy. (ErrorCode: CannotCreateTypeInSealedMetamodel)

If I reference the Ecore element in the `.ecore` via a namespace URI (which makes more sense, to begin with, but is not the default) the MWE2 workflow works, but the Editor shows the same error message.

**Comment by jGleitz** [comment]
I found a workaround by patching the URIs in the registry. With this class:

```Xtend
class RemapEcoreModel {
	@Accessors
	var String from 
	
	def setTo(String uri) {
		EPackage.Registry.INSTANCE.put(uri, EPackage.Registry.INSTANCE.get(from))	
	}
}
```

and this bit in the MWE2 configuration:

```MWE2
	bean = RemapEcoreModel {
		from = "http://www.eclipse.org/emf/2002/Ecore"
		to = "platform:/resource/org.eclipse.emf.ecore/model/Ecore.ecore"
	}
```

everything works. I attached my whole project for reference: [org.xtext.example.mydsl.zip](https://github.com/vitruv-tools/Vitruv/files/6077269/org.xtext.example.mydsl.zip)

**Comment by HeikoKlare** [comment]
That's great, thanks! I'll try that patch.

> 3. We could find out that the problem is at my side and someone else finds a solution.

It seems to result in this option ;-) Although the "solution" is probably not the intended way how it should work in Xtext.

**Comment by jGleitz** [comment]
> Although the "solution" is probably not the intended way how it should work in Xtext

I don’t know whether there _is_ an intended way. But sure, the ‘solution’ is only a workaround (you could also call it a hack).
